From a8957404027c9bfbafe7cab81266a390b4fe3a36 Mon Sep 17 00:00:00 2001
From: Mikolaj Klikowicz <mklikowicz@antmicro.com>
Date: Tue, 10 Feb 2026 17:02:38 +0100
Subject: [PATCH 28/39] tracing: Add reboot magic

This commit adds an event triggered immediately after initialization of
tracing subsystem to detect when Zephyr is reinitialized after reboot.

Signed-off-by: Mikolaj Klikowicz <mklikowicz@antmicro.com>
---
 subsys/tracing/ctf/tsdl/metadata | 17 +++++++++++++++++
 subsys/tracing/tracing_core.c    | 25 ++++++++++++++++++++++++-
 2 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/subsys/tracing/ctf/tsdl/metadata b/subsys/tracing/ctf/tsdl/metadata
index 48c5f318c1d..ab6fc74ce65 100644
--- a/subsys/tracing/ctf/tsdl/metadata
+++ b/subsys/tracing/ctf/tsdl/metadata
@@ -2696,6 +2696,23 @@ event {
 	};
 };
 
+stream {
+	id = 3;
+	event.header := struct event_header;
+	packet.context := struct {
+		uint16_t packet_size;
+	};
+};
+
+event {
+	stream_id = 3;
+	name = event_reboot;
+	id = 0x01;
+	fields := struct {
+		uint64_t magic;
+	};
+};
+
 typealias integer { size = 5; signed = false; } := uint5_t;
 typealias integer { size = 1; signed = false; } := uint1_t;
 typealias integer { size = 2; signed = false; } := uint2_t;
diff --git a/subsys/tracing/tracing_core.c b/subsys/tracing/tracing_core.c
index d84a04f25a7..19179deb212 100644
--- a/subsys/tracing/tracing_core.c
+++ b/subsys/tracing/tracing_core.c
@@ -41,6 +41,9 @@ static atomic_t tracing_state;
 static atomic_t tracing_packet_drop_num;
 static struct tracing_backend *working_backend;
 
+#define MAGIC_EVENT_STREAM_ID 3
+#define MAGIC_EVENT_CONSTANT  3972516962115860257ULL
+
 #ifdef CONFIG_TRACING_ASYNC
 #define TRACING_THREAD_NAME "tracing_thread"
 
@@ -82,9 +85,28 @@ static void tracing_set_state(enum tracing_state state)
 	atomic_set(&tracing_state, state);
 }
 
-static int tracing_init(void)
+static void tracing_reboot_magic(void)
 {
+#ifdef CONFIG_TRACING_CTF
+	struct __packed {
+		uint16_t stream_id;
+		uint16_t packet_size;
+		uint64_t timestamp;
+		uint16_t type;
+		uint64_t magic;
+	} magic_event;
+
+	magic_event.stream_id = MAGIC_EVENT_STREAM_ID;
+	magic_event.packet_size = sizeof(magic_event) * 8;
+	magic_event.timestamp = 0;
+	magic_event.type = 1;
+	magic_event.magic = MAGIC_EVENT_CONSTANT;
+	tracing_buffer_handle((uint8_t *)&magic_event, sizeof(magic_event));
+#endif
+}
 
+static int tracing_init(void)
+{
 	tracing_buffer_init();
 
 	working_backend = tracing_backend_get(CONFIG_TRACING_BACKEND_NAME);
@@ -102,6 +124,7 @@ static int tracing_init(void)
 	} else {
 		tracing_set_state(TRACING_ENABLE);
 	}
+	tracing_reboot_magic();
 
 #ifdef CONFIG_TRACING_ASYNC
 	k_timer_init(&tracing_thread_timer, tracing_thread_timer_expiry_fn, NULL);
-- 
2.47.3

